pkg/controller/migration/convert/core.go

```diff
@@ -367,15 +367,15 @@ func handleNodeSelectors(c *components, install *operatorv1.Installation) error
 
 	// check typha nodeSelectors
 	if c.typha != nil {
-		// we can migrate typha affinities provided they are a NodeAffinity for Preferred.
 		if aff := c.typha.Spec.Template.Spec.Affinity; aff != nil {
-			if aff.PodAffinity != nil || aff.PodAntiAffinity != nil {
-				return ErrIncompatibleCluster{
-					err:       "pod affinity and antiAffinity not supported for typha deployment",
-					component: ComponentTypha,
-					fix:       "remove the affinity",
-				}
+			if aff.PodAffinity != nil {
+				install.Spec.TyphaDeployment.Spec.Template.Spec.Affinity.PodAffinity = aff.PodAffinity
```

Comment:
Since many of the fields here are pointers this does not work in any cases where the fields haven't already been populated.
I think before the if statement you need something like
```
if install.Spec.TyphaDeployment == nil {
  install.Spec.TyphaDeployment = &operatorv1.TyphaDeployment{}
}
if install.Spec.TyphaDeployment.Spec == nil {
  install.Spec.TyphaDeployment.Spec = &operatorv1.TyphaDeploymentSpec{}
}
if install.Spec.TyphaDeployment.Spec.Template == nil {
  install.Spec.TyphaDeployment.Spec.Template = &operatorv1.TyphaDeploymentPodTemplateSpec{}
}
if install.Spec.TyphaDeployment.Spec.Template.Spec == nil {
  install.Spec.TyphaDeployment.Spec.Template.Spec = &operatorv1.TyphaDeploymentPodSpec{}
}
if install.Spec.TyphaDeployment.Spec.Template.Spec.Affinity == nil {
  install.Spec.TyphaDeployment.Spec.Template.Spec.Affinity = &v1.Affinity{}
}
```
That will ensure all the fields are non-nil and won't overwrite the fields if they are already populated.

Reply:
good catch fixed, in addition with some wrong printouts. thanks!

```
go test ./pkg/controller/migration/convert                                                                       1 ↵  10014  16:48:43
ok  	github.com/tigera/operator/pkg/controller/migration/convert	0.903s
```